octopoes/bits/ask_port_specification/ask_port_specification.py

```diff
@@ -0,0 +1,20 @@
+import json
+from pathlib import Path
+from typing import Dict, Iterator, List
+
+from octopoes.models import OOI
+from octopoes.models.ooi.network import Network
+from octopoes.models.ooi.question import Question
+
+
+def run(
+    input_ooi: Network,
+    additional_oois: List,
+    config: Dict[str, str],
+) -> Iterator[OOI]:
+    network = input_ooi
+
+    with (Path(__file__).parent / "question_schema.json").open() as f:
+        schema = json.load(f)
```

Comment:
It is quite nice to get the feedback of the changes without restarting Octopoes though

Reply:
Yes agreed, but it's unlikely that this file be altered during development or production time right? And if it's changed during development, restarting shouldn't take that long.

Reply:
Loading the file also doesn't take too long of course. And if there are issues with the file, perhaps we prefer a bit runtime error instead of an import-time error?

Reply:
I prefer not to do unnecessary IO calls for things that never change. Changing bit (or plugins in general) code or input should always yield a newer version of that plugin.

Runtime errors are worse because they're much harder to catch. In case of the import-time error, that would mean someone developing an invalid plugin build which of course will be caught during a review.

Reply:
Shoudn't we just put the schema in the questionOOI? As this would allow us to also support other questions in our next iterations of this functionality.

Reply:
That QuestionOOI would then in turn need to be created by another bit which we can run on each 'network' being created. The flow would then be:
Bit: For each network, there must be a Ports configOOI, of not, create a QuestionOOI with the jsonschema from this PR. The Form is rendered using the schema value from this QuestionOOI.
The input on this form is then ingested by a normlizer specific for this schema, and creates the configOOI, which then also signals the question to be removed.

Bit: The config is used by the OpenPorts bit(s) to see if there are any issues.

Reply:
> Shoudn't we just put the schema in the questionOOI? As this would allow us to also support other questions in our next iterations of this functionality.

@underdarknl We are putting the schema in the QuestionOOI? We can just add new bits that parse different schema files or parse other schema files in this bit if we want to?


> That QuestionOOI would then in turn need to be created by another bit which we can run on each 'network' being created. The flow would then be: Bit: For each network, there must be a Ports configOOI, of not, create a QuestionOOI with the jsonschema from this PR. The Form is rendered using the schema value from this QuestionOOI. The input on this form is then ingested by a normlizer specific for this schema, and creates the configOOI, which then also signals the question to be removed.
> 
> Bit: The config is used by the OpenPorts bit(s) to see if there are any issues.

I am not sure I understand. The bits don't do checks such as "if X does not exist yet", instead they are mostly idempotent and can be rerun manually in the interface as well. The current QuestionOOI natural key means that is not an issue. The flow you describe is visualized in the ticket and implemented in the subsequent PRs? The only difference now is that I created a generic normalizer for now.

Reply:
The idea of loading schema's for future use is not relevant if you ask me, Its like asking us if we want to pre-load the values of other OOI's too.